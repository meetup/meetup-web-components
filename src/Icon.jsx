import React from 'react';
import cx from 'classnames';

export const ICON_CLASS = 'svg';

export const MEDIA_SIZES = {
	xs: '16',
	s: '24',
	m: '36',
	l: '48',
	xl: '72',
};

const spritePath = require('file!swarm-icons/dist/sprite/sprite.inc');

/**
 * Icon component used to insert an svg icon into a component or page
 *
 * **Accessibility** If an Icon is used on its own without supporting
 * text to explain what it is/does, be a good citizen and pass in an
 * `aria-label` attribute describing what the icon represents
 *
 * @module Icon
 */
class Icon extends React.Component {

	render() {
		const {
			className,
			shape,
			size,
			...other
		} = this.props;

		const classNames = cx(
			ICON_CLASS,
			`svg--${shape}`,
			className
		);

		const dim = MEDIA_SIZES[size];

		return (
			<span className={classNames}>
				<svg
					preserveAspectRatio='xMinYMin meet'
					width={dim}
					height={dim}
					viewBox={`0 0 ${dim} ${dim}`}
					className='svg-icon valign--middle'
					role='img'
					aria-labelledby={`${shape}_title`}
					{...other}>
					<title id={`${shape}_title`} />
					<use role='presentation' xlinkHref={`${spritePath}#icon-${shape}`} />
				</svg>
			</span>
		);
	}
}

Icon.defaultProps = {
	size: 'm'
};

Icon.propTypes = {
	shape: React.PropTypes.string.isRequired,
	size: React.PropTypes.oneOf(Object.keys(MEDIA_SIZES))
};

export default Icon;
