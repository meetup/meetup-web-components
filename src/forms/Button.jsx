import React from 'react';
import cx from 'classnames';
import Flex from '../layout/Flex';
import FlexItem from '../layout/FlexItem';

export const BUTTON_CLASS = 'button';

/**
 * SQ2 Button component
 * @see {@link https://github.com/meetup/sassquatch2/blob/develop/sass/ui-components/_buttons.scss}
 * @see {@link http://meetup.github.io/sassquatch2/ui_components.html#buttons}
 * @module Button
 */
class Button extends React.Component {

	render() {
		const {
			children,
			className,
			onClick,
			reset,
			contrast,
			fullWidth,
			icon,
			primary,
			right,
			small,
			...other
		} = this.props;

		const classNames = cx(
			BUTTON_CLASS,
			{
				'button--contrast': contrast,
				'button--fullWidth': fullWidth,
				'button--primary': primary,
				'button--small': small,
				'button--reset': reset,
			},
			className
		);
		const opts = right ? { rowReverse: 'all' } : {};

		const iconChildren = (
			<Flex className='button--icon-wrapper' {...opts}>
				{icon &&
					<FlexItem shrink className='button--icon valign--middle'>
						{icon}
					</FlexItem>
				}
				{children &&
					<FlexItem className='button--label valign--middle align--center atMedium_align--left'>
						{children}
					</FlexItem>
				}
			</Flex>
		);

		return (
			<button
				className={classNames}
				onClick={onClick}
				role='button'
				{...other}
			>
				{ icon ? iconChildren : children }
			</button>
		);
	}
}

Button.defaultProps = {
	right: false
};

Button.propTypes = {
	reset: React.PropTypes.bool,
	contrast: React.PropTypes.bool,
	fullWidth: React.PropTypes.bool,
	primary: React.PropTypes.bool,
	small: React.PropTypes.bool,
	icon: React.PropTypes.any,
	right: React.PropTypes.bool,
};
export default Button;
