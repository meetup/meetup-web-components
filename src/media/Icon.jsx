import React from 'react';
import cx from 'classnames';
import {
	MEDIA_SIZES,
	MEDIA_QUERIES,
	BREAKPOINT_MEDIA_SCALE_RATIOS
} from '../utils/designConstants';

export const ICON_CLASS = 'svg';

/**
 * Returns a multiplier to use for scaling an Icon based on viewport size
 *
 * @param {Object} mediaMatches - `matches` boolean values by `is${breakpoint}` key
 * @returns {Number} - scale factor
 */
export const getScaleFactor = mediaMatches => {
	let scaleFactor = 1;

	if (mediaMatches.isMedium && mediaMatches.isMedium) {
		scaleFactor = BREAKPOINT_MEDIA_SCALE_RATIOS.medium;
	}
	if (mediaMatches.isLarge && mediaMatches.isLarge) {
		scaleFactor = BREAKPOINT_MEDIA_SCALE_RATIOS.large;
	}

	return scaleFactor;
};

/**
 * Icon component used to insert an svg icon into a component or page
 *
 * **Accessibility** If an Icon is used on its own without supporting
 * text to explain what it is/does, be a good citizen and pass in an
 * `aria-label` attribute describing what the icon represents
 *
 * @module Icon
 */
class Icon extends React.PureComponent {
	constructor(props){
		super(props);
		this.handleMediaChange = this.handleMediaChange.bind(this);

		this.state = {
			iconScaleFactor: 1
		};
	}

	handleMediaChange() {
		if (!this.state.mediaQueries) {
			return;
		}

		this.setState((state) => ({
			iconScaleFactor: getScaleFactor({
				isMedium: state.mediaQueries.medium.matches,
				isLarge: state.mediaQueries.large.matches
			})
		}));
	}

	componentDidMount() {
		if (window.matchMedia) {
			this.setState(() => {
				const mediaQueries = {
					medium: window.matchMedia(MEDIA_QUERIES.medium),
					large: window.matchMedia(MEDIA_QUERIES.large),
				};

				Object.keys(mediaQueries).forEach(mq =>
					mediaQueries[mq].addListener(this.handleMediaChange)
				);

				return {
					mediaQueries
				};
			});
			this.handleMediaChange();
		}
	}

	componentWillUnmount() {
		Object.keys(this.state.mediaQueries).forEach(mq =>
			this.state.mediaQueries[mq].removeListener(this.handleMediaChange)
		);
	}

	render() {
		const {
			className,
			shape,
			size,
			...other
		} = this.props;

		const classNames = cx(
			ICON_CLASS,
			`svg--${shape}`,
			className
		);

		const viewBox = size === 'auto' ? MEDIA_SIZES['xl'] : MEDIA_SIZES[size];
		const dim = Math.floor(MEDIA_SIZES[size] * this.state.iconScaleFactor);

		return (
			<span className={classNames}>
				<svg
					preserveAspectRatio='xMinYMin meet'
					width={dim}
					height={dim}
					viewBox={`0 0 ${viewBox} ${viewBox}`}
					className='svg-icon valign--middle'
					role='img'
					{...other}>
					<use xlinkHref={`#icon-${shape}`}></use>
				</svg>
			</span>
		);
	}
}

Icon.defaultProps = {
	size: 's'
};

Icon.propTypes = {
	shape: React.PropTypes.string.isRequired,
	size: React.PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl', 'auto'])
};

export default Icon;
