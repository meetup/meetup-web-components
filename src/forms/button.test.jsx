import React from 'react';
import { shallow, mount } from 'enzyme';
import Link from 'react-router-dom/Link';
import Button from './Button';
import Icon from '../media/Icon';

describe('Button', () => {
	describe('is a HTML button element', () => {
		it('exists', () => {
			const button = shallow(<Button />);
			expect(button).toMatchSnapshot();
		});
	});

	it('has a default state', () => {
		const button = mount(<Button>Click Me</Button>);
		expect(
			button.getDOMNode().attributes.getNamedItem('data-swarm-button').value
		).toBe('default');
	});

	it('has a primary state', () => {
		const button = mount(<Button primary>Click Me</Button>);
		expect(
			button.getDOMNode().attributes.getNamedItem('data-swarm-button').value
		).toBe('primary');
	});

	it('has a neutral state', () => {
		const button = mount(<Button neutral>Click Me</Button>);
		expect(
			button.getDOMNode().attributes.getNamedItem('data-swarm-button').value
		).toBe('neutral');
	});

	it('has a bordered state', () => {
		const button = mount(<Button bordered>Click Me</Button>);
		expect(
			button.getDOMNode().attributes.getNamedItem('data-swarm-button').value
		).toBe('bordered');
	});

	it('executes onClick when clicked', () => {
		const onClick = jest.fn();
		const button = shallow(<Button onClick={onClick} />);

		button.simulate('click');
		expect(onClick).toHaveBeenCalled();
	});

	describe('Button with icon', () => {
		describe('right aligned', () => {
			it('should set icon container to reverse', () => {
				const icon = <Icon shape="chevron-right" />;
				const button = mount(
					<Button icon={icon} primary right>
						Click me
					</Button>
				);

				expect(
					button.getDOMNode().attributes.getNamedItem('data-icon').value
				).toBe('right');
			});
		});
		describe('right aligned icon on Link', () => {
			it('should set icon container to reverse', () => {
				const icon = <Icon shape="chevron-right" />;
				const button = mount(
					<Button icon={icon} component={Link} primary right>
						Click me
					</Button>
				);

				expect(
					button.getDOMNode().attributes.getNamedItem('data-icon').value
				).toBe('right');
			});
		});
	});

	describe('wrapper component prop', () => {
		const link = 'https://meetup.com/';
		const buttonTagComponent = shallow(<Button href={link}>Button label</Button>);

		it('should render the correct `href` value for anchor tag', () => {
			expect(buttonTagComponent.prop('href')).toBe(link);
		});

		it('should render the correct `to` value for <Link> component', () => {
			const buttonLinkComponent = shallow(
				<Button to={link} component={Link}>
					Button label
				</Button>
			);

			expect(buttonLinkComponent.prop('to')).toBe(link);
		});
	});

	describe('deprecation error', () => {
		it('should error when a Link is passed as a component', () => {
			global.console = { error: jest.fn() };
			shallow(<Button component="div">Click</Button>);
			expect(console.error).toBeCalled();
		});
	});
});
