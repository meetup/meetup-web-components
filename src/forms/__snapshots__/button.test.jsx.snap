// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button is a HTML button element exists 1`] = `
<button
  data-icon="only"
  data-swarm-button="default"
  data-swarm-size="default"
  data-swarm-width="default"
  type="button"
/>
`;
